File: neetcode150/graphs/leetcode130.py
from typing import List


class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.

        Two cells are connected if they are adjacent cells connected horizontally or vertically.

        Surrounded regions should not be on the border, which means that any 'O' on the border of the board
        are not flipped to 'X'. Any 'O' that is not on the border, and it is not connected to an 'O' on the
        border will be flipped to 'X'.

        What we're doing here is we're masking
        """
        rows, cols = len(board), len(board[0])

        if rows == 0:
            return

        for i in range(rows):
            if board[i][0] == "O":
                self.dfs(board, i, 0)

            if board[i][cols - 1] == "O":
                self.dfs(board, i, cols - 1)

        for i in range(cols):
            if board[0][i] == "O":
                self.dfs(board, 0, i)

            if board[rows - 1][i] == "O":
                self.dfs(board, rows - 1, i)

        for i in range(rows):
            for j in range(cols):
                if board[i][j] == "O":
                    board[i][j] = "X"
                elif board[i][j] == "#":
                    board[i][j] = "O"

    def dfs(self, board: List[List[str]], i: int, j: int) -> None:
        rows, cols = len(board), len(board[0])

        if i < 0 or j < 0 or i >= rows or j >= cols or board[i][j] != "O":
            return

        board[i][j] = "#"

        self.dfs(board, i + 1, j)
        self.dfs(board, i - 1, j)
        self.dfs(board, i, j + 1)
        self.dfs(board, i, j - 1)

File: js_algorithms/neetcode150/graphs/leetcode130.js
/**
 * @param {string[][]} board
 * @return {void} Do not return anything, modify board in-place instead.
 */
const solve = function (board) {
  // we should find a value 1 that is not
  // on the border and then do a DFS
  // as long as that value is not on the border
  const rows = board.length;
  if (rows === 0) {
    return;
  }

  const cols = board[0].length;

  // if there exists any Os on the boundary
  // we want to do a DFS and mark any connected Os with a 3rd character

  // go down first and last columns
  for (let i = 0; i < rows; i++) {
    if (board[i][0] === "O") {
      dfs(board, i, 0);
    }

    if (board[i][cols - 1] === "O") {
      dfs(board, i, cols - 1);
    }
  }

  // go over first and last rows
  for (let i = 0; i < cols; i++) {
    if (board[0][i] === "O") {
      dfs(board, 0, i);
    }

    if (board[rows - 1][i] === "O") {
      dfs(board, rows - 1, i);
    }
  }

  // after we've marked those, we should flip those with the 3rd character
  // back to the O and the ones that are still Os into Xs since those are
  // going to be legal flips
  for (let i = 0; i < rows; i++) {
    for (let j = 0; j < cols; j++) {
      if (board[i][j] === "#") {
        board[i][j] = "O";
      } else if (board[i][j] === "O") {
        board[i][j] = "X";
      }
    }
  }
};

/**
 * @param {string[][]} board
 * @param {number} i
 * @param {number} j
 */
function dfs(board, i, j) {
  if (
    i < 0 ||
    j < 0 ||
    i >= board.length ||
    j >= board[0].length ||
    board[i][j] !== "O"
  ) {
    return;
  }

  board[i][j] = "#";

  dfs(board, i + 1, j);
  dfs(board, i - 1, j);
  dfs(board, i, j + 1);
  dfs(board, i, j - 1);
}
